### PR TITLE
sweep: add wallet inputs to reach dust limit

### PR DESCRIPTION
**Pull request by joostjager**
This PR enables the sweeper to sweep inputs that are positively-yielding, but by itself wouldn't reach the dust limit for the sweep transaction output value at the chosen fee rate.

They way to deal with this is to attach additional wallet input(s) to bring up the output value of the transaction.

This is a preparation for sweeping anchor outputs that are too small to reach the dust limit of the sweep tx output.

There are also a few commits in this PR that aren't strictly necessary for the functionality that this PR adds, but are included anyway to prevent touching the code again in the follow-up.

Next step after this PR is to add forced sweeping of negatively yielding inputs.

### CODE REVIEW

File: sweep/sweeper.go [hunk 1]
```diff
@@ -70,16 +70,15 @@ type Params struct {
 }
 
 // pendingInput is created when an input reaches the main loop for the first
-// time. It tracks all relevant state that is needed for sweeping.
+// time. It wraps the input and tracks all relevant state that is needed for
+// sweeping.
 type pendingInput struct {
+	input.Input
```
Comment (halseth) [inline sweep/sweeper.go]: why this change exactly?

File: sweep/tx_input_set.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+// txInputSet is an object that accumulates tx inputs and keeps running counters
+// on various properties of the tx.
+type txInputSet struct {
+	// weightEstimate is the (worst case) tx weight with the current set of
+	// inputs.
+	weightEstimate input.TxWeightEstimator
+
+	// inputTotal is the total value of all inputs.
+	inputTotal btcutil.Amount
```
Comment (halseth) [inline sweep/tx_input_set.go]: can just loop over current inputs to get this instead of storing it as a var?

File: sweep/sweeper.go [hunk 3]
```diff
@@ -789,7 +796,7 @@ func (s *UtxoSweeper) getInputLists(cluster inputCluster,
 	// contain inputs that failed before. Therefore we also add sets
 	// consisting of only new inputs to the list, to make sure that new
 	// inputs are given a good, isolated chance of being published.
-	var newInputs, retryInputs []input.Input
+	var newInputs, retryInputs []txInput
```
Comment (halseth) [inline sweep/sweeper.go]: ~~any reason not using pendingInput directly?~~
got the answer in later commit

File: sweep/tx_input_set.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	weight := b.weightEstimate.Weight()
+	fee := b.feePerKW.FeeForWeight(int64(weight))
+
+	// Calculate the output value if the current input would be
+	// added to the set.
+	newOutputValue := newInputTotal - fee
+
+	// If adding this input makes the total output value of the set
+	// decrease, this is a negative yield input. It shouldn't be
+	// added to the set. We return the current index as the number
```
Comment (halseth) [inline sweep/tx_input_set.go]: comment slightly out of date

File: sweep/txgenerator.go [hunk 5]
```diff
@@ -75,96 +82,74 @@ func generateInputPartitionings(sweepableInputs []input.Input,
...
 // minimizing any negative externalities we cause for the Bitcoin system as a
 // whole.
-func getPositiveYieldInputs(sweepableInputs []input.Input, maxInputs int,
-	feePerKW chainfee.SatPerKWeight) (int, btcutil.Amount) {
-
-	var weightEstimate input.TxWeightEstimator
-
-	// Add the sweep tx output to the weight estimate.
-	weightEstimate.AddP2WKHOutput()
+func addPositiveYieldInputs(inputSet *txInputSet, sweepableInputs []txInput,
```
Comment (halseth) [inline sweep/txgenerator.go]: make this a method on `inputSet`intead?

File: sweep/txgenerator.go [hunk 6]
```diff
@@ -75,96 +82,74 @@ func generateInputPartitionings(sweepableInputs []input.Input,
...
 		}
 
 		log.Infof("Candidate sweep set of size=%v, has yield=%v",
-			count, outputValue)
+			inputCount, txInputs.outputValue)
 
-		sets = append(sets, sweepableInputs[:count])
-		sweepableInputs = sweepableInputs[count:]
+		sweepSet := make([]input.Input, len(txInputs.inputs))
+		copy(sweepSet, txInputs.inputs)
```
Comment (halseth) [inline sweep/txgenerator.go]: why copy here rather than slicing as earlier?

File: sweep/interface.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package sweep
+
+import (
+	"github.com/btcsuite/btcd/wire"
+)
+
+// Wallet contains all wallet related functionality required by sweeper.
+type Wallet interface {
```
Comment (halseth) [inline sweep/interface.go]: 👍

File: sweep/sweeper.go [hunk 8]
```diff
@@ -632,22 +632,24 @@ func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch) {
...
-	// lists of inputs.
-	inputLists, err := s.getInputLists(cluster, currentHeight)
-	if err != nil {
-		return fmt.Errorf("unable to examine pending inputs: %v", err)
-	}
-
-	// Sweep selected inputs.
-	for _, inputs := range inputLists {
-		err := s.sweep(inputs, cluster.sweepFeeRate, currentHeight)
+	return s.cfg.Wallet.WithCoinSelectLock(func() error {
```
Comment (halseth) [inline sweep/sweeper.go]: should add a comment why coin select lock is needed here

File: sweep/txgenerator.go [hunk 9]
```diff
@@ -105,13 +107,22 @@ func generateInputPartitionings(sweepableInputs []txInput,
 		// continuing with the remaining inputs will only lead to sets
 		// with a even lower output value.
 		if txInputs.outputValue < dustLimit {
-			log.Debugf("Set value %v below dust limit of %v",
-				txInputs.outputValue, dustLimit)
-			return sets, nil
+			dustLimitReached, err := addWalletInputs(
```
Comment (halseth) [inline sweep/txgenerator.go]: should have comment

File: sweep/txgenerator.go [hunk 10]
```diff
@@ -123,6 +134,87 @@ func generateInputPartitionings(sweepableInputs []txInput,
...
 }
 
+// addWalletInputs retrieves utxos from the wallet and tries adding as many as
+// required to bring the tx output value above the given minimum.
+func addWalletInputs(inputSet *txInputSet, minOutputAmt btcutil.Amount,
+	wallet Wallet) (bool, error) {
+
+	// Retrieve wallet utxos. Only consider confirmed utxos to prevent
+	// problems around RBF rules for unconfirmed inputs.
+	utxos, err := wallet.ListUnspentWitness(1, math.MaxInt32)
```
Comment (halseth) [inline sweep/txgenerator.go]: according to the godoc minConfs can be set `=0` here? I'm not sure what is correct or if it's the same, worth checking out probably.

File: sweep/txgenerator.go [hunk 11]
```diff
@@ -123,6 +134,87 @@ func generateInputPartitionings(sweepableInputs []txInput,
...
+// to the other inputs to sweep.
+func createWalletTxInput(utxo *lnwallet.Utxo) (txInput, error) {
+	var witnessType input.WitnessType
+	switch utxo.AddressType {
+	case lnwallet.WitnessPubKey:
+		witnessType = input.WitnessKeyHash
+	case lnwallet.NestedWitnessPubKey:
+		witnessType = input.NestedWitnessKeyHash
+	default:
+		return nil, fmt.Errorf("unknown input witness")
```
Comment (halseth) [inline sweep/txgenerator.go]: add the unkown type to error

File: sweep/tx_input_set.go [hunk 12]
```diff
@@ -71,6 +74,20 @@ func (b *txInputSet) add(input txInput) bool {
 		return false
 	}
 
+	// If this input comes from the wallet, verify that we still gain
+	// something with this transaction.
+	_, fromWallet := input.(*walletTxInput)
+	if fromWallet {
+		newWalletTotal := b.walletInputTotal + value
```
Comment (halseth) [inline sweep/tx_input_set.go]: taken in isolation, it looks like this can make the effective yield go down. Zooming out it should be okay, since we only add this when we are below the dustlimit, which obviously is what we want, even though it effectively costs us more because of the extra input. But this is not obvious from reading it in isolation.

Some comments would go a long way making it more clear I think.

File: sweep/txgenerator.go [hunk 13]
```diff
@@ -123,6 +134,87 @@ func generateInputPartitionings(sweepableInputs []txInput,
...
+	)
+
+	return &walletTxInput{
+		Input: walletInput,
+	}, nil
+}
+
+// walletTxInput wraps a wallet input in the same way other sweeper inputs are
+// wrapped.
+type walletTxInput struct {
```
Comment (halseth) [inline sweep/txgenerator.go]: instead of having  a special `walletTxInput` type, what do you think about having the `txInput` define a `gain` or similar. It would be `0` for wallet inputs, and the actual  output value for other types we sweep.

When adding the wallet input, it would check that the total `gain-fee` is positive.

File: sweep/sweeper.go (hunk 1 above)
Comment (joostjager) [reply sweep/sweeper.go]: Two reasons:
1. Less code
2. Allows us to use `pendingInput` as `txInput` in next commits.

File: sweep/tx_input_set.go (hunk 2 above)
Comment (joostjager) [reply sweep/tx_input_set.go]: This struct already tracks several running counters. Why would we recalculate `inputTotal` specifically every time we need it?

File: sweep/tx_input_set.go (hunk 4 above)
Comment (joostjager) [reply sweep/tx_input_set.go]: fixed

File: sweep/txgenerator.go (hunk 6 above)
Comment (joostjager) [reply sweep/txgenerator.go]: A suggestion from the linter. I like it because it is less code. Apparently this is the way to do covariance.

File: sweep/txgenerator.go (hunk 5 above)
Comment (joostjager) [reply sweep/txgenerator.go]: good one, moved

File: sweep/sweeper.go (hunk 3 above)
Comment (joostjager) [reply sweep/sweeper.go]: the main reason is a bit of isolation from everything else that the sweeper tracks.

File: sweep/sweeper.go (hunk 8 above)
Comment (joostjager) [reply sweep/sweeper.go]: added

File: sweep/txgenerator.go (hunk 9 above)
Comment (joostjager) [reply sweep/txgenerator.go]: added

File: sweep/txgenerator.go (hunk 10 above)
Comment (joostjager) [reply sweep/txgenerator.go]: Where did you see this? I find in btcwallet the following:

```
// Outputs with fewer confirmations than the minimum or more
// confs than the maximum are excluded.
```

I'd think we need to set min to 1?

File: sweep/txgenerator.go (hunk 11 above)
Comment (joostjager) [reply sweep/txgenerator.go]: fixed

File: sweep/txgenerator.go (hunk 13 above)
Comment (joostjager) [reply sweep/txgenerator.go]: I still need to create an object that conforms to the `txInput` interface. I don't want to use `pendingInput` for this.

Comment (joostjager) [reply sweep/txgenerator.go]: Gain could also work, but it would then always be either zero or the input value. A bool would do the same thing. But adding a bool also seemed redundant if I can just type check here.

File: sweep/tx_input_set.go (hunk 12 above)
Comment (joostjager) [reply sweep/tx_input_set.go]: Comments added

File: sweep/txgenerator.go (hunk 13 above)
Comment (joostjager) [reply sweep/txgenerator.go]: Decided to make a `fromWallet` boolean for now.

**Comment by joostjager** [comment]
Unit test added

File: sweep/tx_input_set.go [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+		// Try to add the input to the transaction. If that doesn't
+		// succeed because it wouldn't increase the output value,
+		// return. Assuming inputs are sorted by yield, any further
+		// inputs wouldn't increase the output value either.
+		if !b.add(input) {
+			return
+		}
+
+		// Stop if max inputs is reached.
+		if idx == maxInputs-1 {
```
Comment (wpaulino) [inline sweep/tx_input_set.go]: Shouldn't `maxInputs` be enforced to `b.inputs` if `addPositiveYieldInputs` is called multiple times to the same set?

File: sweep/tx_input_set.go [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+	// Add the sweep tx output to the weight estimate.
+	b.weightEstimate.AddP2WKHOutput()
+
+	return &b
+}
+
+// dustLimitReached returns true if we've accumulated enough inputs to meet the
+// dust limit.
+func (b *txInputSet) dustLimitReached() bool {
```
Comment (wpaulino) [inline sweep/tx_input_set.go]: Nit: what does `b` stand for here?

File: sweep/tx_input_set.go [hunk 16]
```diff
@@ -122,3 +171,71 @@ func (b *txInputSet) addPositiveYieldInputs(sweepableInputs []txInput,
...
+
+	// Retrieve wallet utxos. Only consider confirmed utxos to prevent
+	// problems around RBF rules for unconfirmed inputs.
+	utxos, err := b.wallet.ListUnspentWitness(1, math.MaxInt32)
+	if err != nil {
+		return err
+	}
+
+	for _, utxo := range utxos {
+		input, err := createWalletTxInput(utxo)
```
Comment (wpaulino) [inline sweep/tx_input_set.go]: Don't we want to lock these if they get chosen?

File: sweep/sweeper.go [hunk 17]
```diff
@@ -719,6 +724,8 @@ func (s *UtxoSweeper) scheduleSweep(currentHeight int32) error {
 	startTimer := false
 	for _, cluster := range s.clusterBySweepFeeRate() {
 		// Examine pending inputs and try to construct lists of inputs.
+		// We don't need to obtain the coin selection lock, because
+		// those coins may not be available when we sweep anyway.
```
Comment (wpaulino) [inline sweep/sweeper.go]: May not be available because they were used in another transaction?

File: sweep/sweeper.go [hunk 18]
```diff
@@ -640,8 +639,7 @@ func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch) {
 func (s *UtxoSweeper) bucketForFeeRate(
 	feeRate chainfee.SatPerKWeight) int {
 
-	minBucket := s.relayFeeRate + chainfee.SatPerKWeight(s.cfg.FeeRateBucketSize)
```
[the remark this answers is not in the document]
Comment (joostjager) [reply sweep/sweeper.go]: The first bucket still starts at the min relay fee. But previously the bucket size was `relayFeeRate + bucketSize`. I think that isn't what one would expect when you set a bucket size?

File: sweep/tx_input_set.go (hunk 15 above)
Comment (joostjager) [reply sweep/tx_input_set.go]: converted to a nice `t`

File: sweep/tx_input_set.go [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+// input was added to the set. An input is rejected if it decreases the tx
+// output value after paying fees.
+func (b *txInputSet) add(input input.Input) bool {
+	// Can ignore error, because it has already been checked when
+	// calculating the yields.
+	size, isNestedP2SH, _ := input.WitnessType().SizeUpperBound()
+
+	// Keep a running weight estimate of the input set.
+	if isNestedP2SH {
+		b.weightEstimate.AddNestedP2WSHInput(size)
```
[the remark this answers is not in the document]
Comment (joostjager) [reply sweep/tx_input_set.go]: Very good catch. It wasn't a problem previously, but because we are now trying another add after a failed add, it has become important. Made a copy now.

File: sweep/tx_input_set.go (hunk 14 above)
Comment (joostjager) [reply sweep/tx_input_set.go]: That doesn't happen at the moment, but changed this just in case.

File: sweep/sweeper.go (hunk 17 above)
Comment (joostjager) [reply sweep/sweeper.go]: Rewritten comment. The basic idea is that we are just after an indication here

File: sweep/tx_input_set.go (hunk 16 above)
Comment (joostjager) [reply sweep/tx_input_set.go]: If we are actually sweeping, this is executed within a coin selection lock. So no other processes should be selecting coins then.

File: sweep/tx_input_set.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply sweep/tx_input_set.go]: Unit test added for this

File: sweep/txgenerator.go (hunk 6 above)
Comment (halseth) [reply sweep/txgenerator.go]: Looks like it is more code tho? Anyway, no big deal, linter is always right.

File: sweep/sweeper.go [hunk 20]
```diff
@@ -645,6 +628,28 @@ func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch) {
 	}
 }
 
+// sweepCluster tries to sweep the given input cluster.
+func (s *UtxoSweeper) sweepCluster(cluster inputCluster,
+	currentHeight int32) error {
+
+	// Examine pending inputs and try to construct
```
Comment (halseth) [inline sweep/sweeper.go]: nit: comment fits one one line.

File: sweep/tx_input_set.go [hunk 21]
```diff
@@ -125,3 +174,71 @@ func (t *txInputSet) addPositiveYieldInputs(sweepableInputs []txInput,
...
+		}
+
+		// Return if we've reached the minimum output amount.
+		if t.dustLimitReached() {
+			return nil
+		}
+	}
+
+	// We were not able to reach the minimum output amount.
+	return nil
```
Comment (halseth) [inline sweep/tx_input_set.go]: nit: can return error and remove the dustlimit check after calling this method.

File: sweep/tx_input_set.go [hunk 22]
```diff
@@ -125,3 +174,71 @@ func (t *txInputSet) addPositiveYieldInputs(sweepableInputs []txInput,
...
+// many as required to bring the tx output value above the given minimum.
+func (t *txInputSet) tryAddWalletInputsIfNeeded() error {
+	// If we've already reached the dust limit, no action is needed.
+	if t.dustLimitReached() {
+		return nil
+	}
+
+	// Retrieve wallet utxos. Only consider confirmed utxos to prevent
+	// problems around RBF rules for unconfirmed inputs.
+	utxos, err := t.wallet.ListUnspentWitness(1, math.MaxInt32)
```
Comment (halseth) [inline sweep/tx_input_set.go]: to avoid having the wallet as a field on the input set, the list of utxos can be passed into this method.

File: sweep/interface.go [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+	// PublishTransaction performs cursory validation (dust checks, etc) and
+	// broadcasts the passed transaction to the Bitcoin network.
+	PublishTransaction(tx *wire.MsgTx) error
+
+	// ListUnspentWitness returns all unspent outputs which are version 0
+	// witness programs. The 'minconfirms' and 'maxconfirms' parameters
+	// indicate the minimum and maximum number of confirmations an output
+	// needs in order to be returned by this method. Passing -1 as
```
Comment (halseth) [inline sweep/interface.go]: Here is the confusing minConf comment. Does passing 0 mean unconfirmed or confirmed outputs? 😛 

Probably best to pass 1, and just test that actually works.

File: sweep/txgenerator.go (hunk 10 above)
Comment (halseth) [reply sweep/txgenerator.go]: See other comment. But yeah, setting to 1 is probably correct.

File: sweep/tx_input_set.go (hunk 12 above)
Comment (halseth) [reply sweep/tx_input_set.go]: Nice, easier to read now :)

File: sweep/tx_input_set.go (hunk 2 above)
Comment (joostjager) [reply sweep/tx_input_set.go]: Resolving this comment because it is the same as the one below

File: sweep/tx_input_set.go [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+	// the set and return the outcome.
+	if newOutputValue <= t.outputValue {
+		return false
+	}
+
+	// Update running values.
+	t.inputTotal = newInputTotal
+	t.outputValue = newOutputValue
+	t.inputs = append(t.inputs, input)
+	t.weightEstimate = newWeightEstimate
```
[the remark this answers is not in the document]
Comment (joostjager) [reply sweep/tx_input_set.go]: In a later commit, i would also need the `fromWallet` state. I am not too convinced of it. When you add 20 inputs, 200 (virtual) add operations will be executed. (1+2+3+...+20).

@wpaulino what do you think?

File: sweep/txgenerator.go (hunk 6 above)
Comment (joostjager) [reply sweep/txgenerator.go]: Oh, you mean even earlier. There was an intermediate revision that had a for loop to make the copy instead of calling `copy`. That is what the linter suggested.

But after that some types changed and it is again possible to slice. Fixed

File: sweep/interface.go (hunk 23 above)
Comment (joostjager) [reply sweep/interface.go]: Ah ok. That was copy pasted. I removed the story about `-1`. It seems to be a left over from old code.

File: sweep/tx_input_set.go (hunk 22 above)
Comment (joostjager) [reply sweep/tx_input_set.go]: yes, but then we'd always get utxos, even if the dust limit is already reached. could also move that out then. but not sure if it is better

File: sweep/tx_input_set.go (hunk 21 above)
Comment (joostjager) [reply sweep/tx_input_set.go]: idk, it isn't really an error. i could return an additional bool, but to me it isn't all that different.

File: sweep/tx_input_set_test.go [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+		t.Fatal(err)
+	}
+
+	if !set.dustLimitReached() {
+		t.Fatal("expected dust limit to be reached")
+	}
+}
+
+// testInput returns a test input with the specified amount.
+func testInput(amt btcutil.Amount) input.Input {
```
Comment (wpaulino) [inline sweep/tx_input_set_test.go]: Why not use `createTestInput`?

File: sweep/tx_input_set.go [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+// addPositiveYieldInputs adds sweepableInputs that have a positive yield to the
+// input set. This function assumes that the list of inputs is sorted descending
+// by yield.
+//
+// TODO(roasbeef): Consider including some negative yield inputs too to clean
+// up the utxo set even if it costs us some fees up front.  In the spirit of
+// minimizing any negative externalities we cause for the Bitcoin system as a
+// whole.
+func (t *txInputSet) addPositiveYieldInputs(sweepableInputs []txInput,
+	maxInputs int) {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply sweep/tx_input_set.go]: It is a method? Or what do you mean with 'member'?

[the remark this answers is not in the document]
Comment (joostjager) [reply sweep/tx_input_set.go]: Ah, max inputs. Ok

[the remark this answers is not in the document]
Comment (joostjager) [reply sweep/tx_input_set.go]: Added

File: sweep/sweeper.go [hunk 27]
```diff
@@ -718,6 +724,10 @@ func (s *UtxoSweeper) scheduleSweep(currentHeight int32) error {
 	startTimer := false
 	for _, cluster := range s.clusterBySweepFeeRate() {
 		// Examine pending inputs and try to construct lists of inputs.
+		// We don't need to obtain the coin selection lock, because we
+		// just need an indication as to whether we can sweep. More
+		// inputs may be added until we publish the transaction and
+		// coins that we select now may be used in other transactions.
```
[the remark this answers is not in the document]
Comment (joostjager) [reply sweep/sweeper.go]: We opportunistically schedule a sweep, I think that is how this should be looked at.

File: sweep/tx_input_set_test.go (hunk 25 above)
Comment (joostjager) [reply sweep/tx_input_set_test.go]: Ok, made a wrapper for the existing `createTestInput`
